core/src/main/java/cz/o2/proxima/storage/watermark/GlobalWatermarkTracker.java

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright 2017-2020 O2 Czech Republic, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.storage.watermark;
+
+import cz.o2.proxima.annotations.Evolving;
+import cz.o2.proxima.time.WatermarkSupplier;
+import java.time.Instant;
+import java.util.Map;
+
+/**
+ * A tracker of global progress among multiple (distributed) processes.
+ *
+ * <p>A {@link GlobalWatermarkTracker} consists of the following:
+ * <li>Name of the tracker. The name MUST uniquely identify the tracker among all possible global
+ *     trackers. The name also serves as grouping identifier for calculation of the value of global
+ *     watermark.
+ * <li>Predefined list of <i>processes</i>.
+ */
+@Evolving
+public interface GlobalWatermarkTracker extends WatermarkSupplier {
+
+  /**
+   * Setup all parallel consumers. This MUST be used once when the tracker is constructed to setup
+   * initial names and watermarks of consumers. Note that consumers might be added during runtime,
+   * but <b>their added watermark cannot move the watermark back in time</b>. That is to say -
+   *
+   * @param processes map of process name to watermark
+   */
+  void setup(Map<String, Instant> processes);
+
+  /**
+   * Update watermark of given process. This call MAY add a new process, which was not part of
+   * {@link #setup(Map)}.
+   *
+   * @param name name of the process
+   * @param currentWatermark current processing watermark of the process
+   */
+  void update(String name, Instant currentWatermark);
+
+  /**
+   * Remove given process from the tracker. The watermark of the process (if any) will no longer
+   * hold the global watermark.
+   *
+   * <p>Note that this is semantically equivalent to call to {@link #update}(name,
+   * Instant.ofEpochMilli(Long.MAX_VALUE))
+   *
+   * @param name name of the process to remove
+   */
+  default void finished(String name) {
```

Comment:
we should rename this to `finish` to keep method names consistent` (eg.: update vs updated)

Reply:
I don't agree that there is any reason for these methods to be exacty aligned. While `update` is a command - i.e. "please update watermark of this consumer to ...", `finished` means "please be notified that there will no longer be any updates from this consumer".